lib/Sema/MiscDiagnostics.cpp

```diff
@@ -3707,7 +3707,9 @@ ASTWalker::PreWalkResult<Expr *> VarDeclUsageChecker::walkToExprPre(Expr *E) {
     return Action::SkipChildren(E);
   }
 
-  assert(AllExprsSeen.insert(E).second && "duplicate traversal");
+  // TODO: `MacroExpansionExpr` may have a substitute `MacroExpansionDecl` with
+  // the same argument expressions. How should we assert here?
```

Comment:
This representation doesn't seem correct to me, because much of the type checker is not setup to visit AST nodes twice. If you're re-using AST nodes, maybe we need to be using `OpaqueValueExpr` or similar for the argument values to avoid re-visiting AST nodes, or maybe CSApply needs to re-write `MacroExpansionExpr` to a more precise representation.

Reply:
`OpaqueValueExpr` worked. Thanks.

> maybe CSApply needs to re-write MacroExpansionExpr to a more precise representation.

The more precise representation would not be an expression anymore but a `MacroExpansionDecl`. CSApply doesn't seem to be able to rewrite exprs as decls.  I think my earlier iteration had a far better representation (`MacroExpansionDecl` expanding to `MacroExpansionExpr`) but that would require significant backtracking or lookahead during parsing to handle the `#macro + 1` case.